rbm_softmax.py

```diff
@@ -0,0 +1,504 @@
+
+        # TODO: I don't have a great intuition about this. Why multiply by n_features?
+        # The overfitting section of "Practical Guide" just says to compare the
+        # average free energy of train and validation and to compare them. Any reason
+        # not to just do that here as well? Seems much simpler to interpret. Maybe
+        # it's because we're supposed to be dealing with smaller deltas here?
+        #return v.shape[1] * log_logistic(fe_corrupted - fe)
+
+    @common.timeit
+    def score_validation_data(self, train, validation):
+        """Return the energy difference between the given validation data, and a
+        subset of the training data. This is useful for monitoring overfitting.
+        If the model isn't overfitting, the difference should be around 0. The
+        greater the difference, the more the model is overfitting.
+        """
+        # It's important to use the same subset of the training data every time (per Hinton's "Practical Guide")
+        return self._free_energy(train[:validation.shape[0]]).mean(), self._free_energy(validation).mean()
+
+    def fit(self, X, validation=None):
+        """Fit the model to the data X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} shape (n_samples, n_features)
+            Training data.
+
+        validation : {array-like, sparse matrix}
+
+        Returns
+        -------
+        self : BernoulliRBM
+            The fitted model.
+        """
+        X = check_array(X, accept_sparse='csr', dtype=np.float)
+        n_samples = X.shape[0]
+
+        if not hasattr(self, 'components_'):
+            self.components_ = np.asarray(
+                self.rng_.normal(0, 0.01, (self.n_components, X.shape[1])),
+                order='fortran')
+            self.intercept_hidden_ = np.zeros(self.n_components, )
+            #self.intercept_visible_ = np.zeros(X.shape[1], )
+            # 'It is usually helpful to initialize the bias of visible unit i to log[p_i/(1-p_i)] where p_i is the prptn of training vectors where i is on' - Practical Guide
+            counts = X.sum(axis=0).A.reshape(-1)
+            # There should be no units that are always on
+            assert np.max(counts) < X.shape[0], "Found a visible unit always on in the training data. Fishy."
+            # There might be some units never on. Add a pseudo-count of 1 to avoid inf
+            vis_priors = (counts + 1) / float(X.shape[0])
+            self.intercept_visible_ = np.log( vis_priors / (1 - vis_priors) )
+
+
+        else:
+            print "Resuing existing weights and biases"
+        # Don't necessarily want to reuse h_samples if we have one leftover from before - batch size might have changed
+        self.h_samples_ = np.zeros((self.batch_size, self.n_components))
+
+        n_batches = int(np.ceil(float(n_samples) / self.batch_size))
+        batch_slices = list(gen_even_slices(n_batches * self.batch_size,
```

Comment:
I just noticed batches aren't shuffled between epochs. But that doesn't really matter - right?